r2/types.go

```diff
@@ -0,0 +1,308 @@
+		}
+		(*f)[parts[0]] = parts[1]
+	}
+	return nil
+}
+
+type mappingRuleJSON struct {
+	ID             string          `json:"id,omitempty"`
+	Name           string          `json:"name" validate:"required"`
+	CutoverMillies int64           `json:"cutoverMillies,omitempty"`
+	Filters        filters         `json:"filter" validate:"required"`
+	Policies       []policy.Policy `json:"policies" validate:"required"`
+}
+
+func (m mappingRuleJSON) mappingRuleView() rules.MappingRuleView {
+	return rules.MappingRuleView{
+		ID:       m.ID,
+		Name:     m.Name,
+		Filters:  m.Filters,
+		Policies: m.Policies,
+	}
+}
+
+func newMappingRuleJSON(mrv *rules.MappingRuleView) mappingRuleJSON {
+	return mappingRuleJSON{
+		ID:             mrv.ID,
+		Name:           mrv.Name,
+		Filters:        mrv.Filters,
+		Policies:       mrv.Policies,
+		CutoverMillies: nanoToMilli(mrv.CutoverNanos),
+	}
+}
+
+type mappingRuleHistory struct { // nolint
+	States []mappingRuleJSON `json:"states"`
+}
+
+// RollupTargetData ...
+type rollupTargetJSON struct {
+	Name     string          `json:"name" validate:"required"`
+	Tags     []string        `json:"tags" validate:"required"`
+	Policies []policy.Policy `json:"policies" validate:"required"`
+}
+
+func (t rollupTargetJSON) rollupTargetView() rules.RollupTargetView {
+	return rules.RollupTargetView{
+		Name:     t.Name,
+		Tags:     t.Tags,
+		Policies: t.Policies,
+	}
+}
+
+func newRollupTargetJSON(t rules.RollupTargetView) rollupTargetJSON {
+	return rollupTargetJSON{
+		Name:     t.Name,
+		Tags:     t.Tags,
+		Policies: t.Policies,
+	}
+}
+
+type rollupRuleJSON struct {
+	ID             string             `json:"id,omitempty"`
+	Name           string             `json:"name" validate:"required"`
+	Filters        filters            `json:"filter" validate:"required"`
+	Targets        []rollupTargetJSON `json:"targets" validate:"required,dive,required"`
+	CutoverMillies int64              `json:"cutoverMillies,omitempty"`
+}
+
+func newRollupRuleJSON(rrv *rules.RollupRuleView) rollupRuleJSON {
+	targets := make([]rollupTargetJSON, len(rrv.Targets))
+	for i, t := range rrv.Targets {
+		targets[i] = newRollupTargetJSON(t)
+	}
+	return rollupRuleJSON{
+		ID:             rrv.ID,
+		Name:           rrv.Name,
+		Filters:        rrv.Filters,
+		Targets:        targets,
+		CutoverMillies: nanoToMilli(rrv.CutoverNanos),
+	}
+}
+
+func (r rollupRuleJSON) rollupRuleView() rules.RollupRuleView {
+	targets := make([]rules.RollupTargetView, len(r.Targets))
+	for i, t := range r.Targets {
+		targets[i] = t.rollupTargetView()
+	}
+
+	return rules.RollupRuleView{
+		ID:      r.ID,
+		Name:    r.Name,
+		Filters: r.Filters,
+		Targets: targets,
+	}
+}
+
+type rollupRuleHistory struct { // nolint
+	States []rollupRuleJSON `json:"states"`
+}
+
+func nanoToMilli(nanos int64) int64 {
+	return nanos / 1000000
+}
+
+// ConflictError represents either a version mismatch writing data or a data conflict issue.
```

Comment:
Do you mean for these to be exported types?

Reply:
Yeah, I'm creating them in the implementations of the store interface.

Reply:
Is the idea for the API user to be able to distinguish different types of failures?

Reply:
More for someone implementing a store to know how to map http codes to errors.